File: DFS of Graph - GFG/dfs-of-graph.py
#User function Template for python3

class Solution:
    
    #Function to return a list containing the DFS traversal of the graph.
    def util(self, v, visited, adj, output):
        visited.add(v)
        output.append(v)
        for i in adj[v]:
            if i not in visited:
                self.util(i, visited, adj, output)
    def dfsOfGraph(self, V, adj):
        # code here
        visited = set()
        output = []
        self.util(0, visited, adj, output)
        return output


#{ 
 # Driver Code Starts
if __name__ == '__main__':
    T=int(input())
    while T>0:
        V,E=map(int,input().split())
        adj=[[] for i in range(V+1)]
        for i in range(E):
            u,v=map(int,input().split())
            adj[u].append(v)
            adj[v].append(u)
        ob=Solution()
        ans=ob.dfsOfGraph(V,adj)
        for i in range(len(ans)):
            print(ans[i],end=" ")
        print()
        T-=1
# } Driver Code Ends

File: DFS of Graph - GFG/dfs-of-graph.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution {
  public:
    void dfs(int V,  unordered_map<int, bool> &visited, vector<int> adj[], vector<int> &ans){
        visited[V] = 1;
        ans.push_back(V);
        for(auto i: adj[V]){
            if(!visited[i]){
                dfs(i, visited, adj, ans);
            }
        }
    }
    // Function to return a list containing the DFS traversal of the graph.
    vector<int> dfsOfGraph(int V, vector<int> adj[]) {
        // Code here
        unordered_map<int, bool> visited;
        vector<int> ans;
        // stack<int> s;
        
        dfs(0, visited, adj, ans);
        return ans;
    }
};

//{ Driver Code Starts.
int main() {
    int tc;
    cin >> tc;
    while (tc--) {
        int V, E;
        cin >> V >> E;

        vector<int> adj[V];

        for (int i = 0; i < E; i++) {
            int u, v;
            cin >> u >> v;
            adj[u].push_back(v);
            adj[v].push_back(u);
        }
        // string s1;
        // cin>>s1;
        Solution obj;
        vector<int> ans = obj.dfsOfGraph(V, adj);
        for (int i = 0; i < ans.size(); i++) {
            cout << ans[i] << " ";
        }
        cout << endl;
    }
    return 0;
}
// } Driver Code Ends